src/sabre_swap.rs

```diff
@@ -0,0 +1,138 @@
+use crate::swap_scores::SwapScores;
+
+const EXTENDED_SET_WEIGHT: f64 = 0.5;
+
+#[pyclass]
+pub enum Heuristic {
+    Basic,
+    Lookahead,
+    Decay,
+}
+
+#[pyfunction]
+pub fn sabre_score_heuristic(
+    layer: EdgeList,
+    layout: &NLayout,
+    swap_scores: &mut SwapScores,
+    extended_set: EdgeList,
+    distance_matrix: PyReadonlyArray2<f64>,
+    qubits_decay: QubitsDecay,
+    heuristic: &Heuristic,
+) -> Vec<[usize; 2]> {
+    let dist = distance_matrix.as_array();
+    swap_scores
+        .scores
+        .par_iter_mut()
+        .for_each(|(swap_qubits, score)| {
+            let mut trial_layout = layout.clone();
+            trial_layout.swap_logic(swap_qubits[0], swap_qubits[1]);
+            *score = score_heuristic(
+                heuristic,
+                &layer.edges,
+                &extended_set.edges,
+                &trial_layout,
+                swap_qubits,
+                &dist,
+                &qubits_decay.decay,
+            );
+        });
+    let min_score = swap_scores
+        .scores
+        .par_values()
+        .min_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap();
+    let mut best_swaps: Vec<[usize; 2]> = swap_scores
+        .scores
+        .iter()
+        .filter_map(|(k, v)| if v == min_score { Some(*k) } else { None })
+        .collect();
+    best_swaps.par_sort();
+    best_swaps
+}
+
+#[inline]
+fn compute_cost(layer: &[[usize; 2]], layout: &NLayout, dist: &ArrayView2<f64>) -> f64 {
+    layer
+        .iter()
+        .map(|gate| dist[[layout.logic_to_phys[gate[0]], layout.logic_to_phys[gate[1]]]])
+        .sum()
+}
+
+fn score_lookahead(
+    layer: &[[usize; 2]],
+    extended_set: &[[usize; 2]],
+    layout: &NLayout,
+    dist: &ArrayView2<f64>,
+) -> f64 {
+    let mut first_cost = compute_cost(layer, layout, dist);
+    first_cost /= layer.len() as f64;
+    let second_cost = if !extended_set.is_empty() {
+        compute_cost(extended_set, layout, dist)
+    } else {
+        0.
+    };
+    first_cost + EXTENDED_SET_WEIGHT * second_cost
+}
+
+fn score_decay(
+    layer: &[[usize; 2]],
+    extended_set: &[[usize; 2]],
+    layout: &NLayout,
+    dist: &ArrayView2<f64>,
+    swap_qubits: &[usize; 2],
+    qubits_decay: &[f64],
+) -> f64 {
+    let total_cost = score_lookahead(layer, extended_set, layout, dist);
+    qubits_decay[swap_qubits[0]].max(qubits_decay[swap_qubits[1]]) * total_cost
+}
+
+fn score_heuristic(
+    heuristic: &Heuristic,
+    layer: &[[usize; 2]],
+    extended_set: &[[usize; 2]],
+    layout: &NLayout,
+    swap_qubits: &[usize; 2],
+    dist: &ArrayView2<f64>,
+    qubits_decay: &[f64],
+) -> f64 {
+    match heuristic {
+        Heuristic::Basic => compute_cost(layer, layout, dist),
+        Heuristic::Lookahead => score_lookahead(layer, extended_set, layout, dist),
+        Heuristic::Decay => {
+            score_decay(layer, extended_set, layout, dist, swap_qubits, qubits_decay)
+        }
+    }
+}
```

Comment:
Does Rust have function pointers / first-class functions?  It feels like it ought to be more efficient to do this switch block one stack frame higher, but maybe the compiler or modern branch prediction is good enough that it doesn't matter, and the different parameters would be a nuisance.

Reply:
You can do that with rust, I think the problem though is the arguments are different for each function so the type checker will complain that `fn(&[[usize; 2]], &NLayout, &ArrayView2<f64>) -> f64` is different from `fn(&[[usize; 2]], &[[usize; 2]], &NLayout, &ArrayView2<f64>) -> f64` and fn(&[[usize; 2]], &[[usize; 2]], &NLayout, &ArrayView2<f64>, &[f64]) -> f64`. To fix this we can add unused arguments to basic and lookahead so they all match. That being said I assume the compiler is probably smart enough to optimize it away and even if it didn't yeah I assume the branch predictor will predict it correctly. Once I get this more finalized and the performance where it should be I can look at doing this as a potential follow on optimization.

Reply:
I need to do some more work with Rust even if only to recalibrate my programming expectations from a pure interpreted language like Python to one that passes through an optimising compiler - it's been a while since I did any major amount of work in a language where you didn't have to do the compiler's job for it!